src/NuGet.Core/NuGet.ProjectModel/Utf8JsonStreamReader.cs

```diff
@@ -0,0 +1,338 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Buffers;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Text;
+using System.Text.Json;
+
+namespace NuGet.ProjectModel
+{
+    /// <summary>
+    /// This struct is used to read over a memeory stream in parts, in order to avoid reading the entire stream into memory.
+    /// It functions as a wrapper around <see cref="Utf8JsonStreamReader"/>, while maintaining a stream and a buffer to read from.
+    /// must call <see cref="CompleteRead"/> to return the buffer to the pool when completed.
+    /// </summary>
+    internal ref struct Utf8JsonStreamReader
+    {
+        private static readonly char[] DelimitedStringDelimiters = new char[] { ' ', ',' };
+
+        // The UT8 BOM characters which may appear at the start of a JSON file.
+        private ReadOnlySpan<byte> _utf8Bom = new byte[] { 0xEF, 0xBB, 0xBF };
+        private Utf8JsonReader _reader;
+        // The buffer is used to read from the stream in chunks.
+        private byte[] _buffer;
+        // The stream is the source of the JSON data.
+        private Stream _stream;
+        private bool _complete;
+
+        internal Utf8JsonStreamReader(Stream stream)
+        {
+            if (stream is null)
+            {
+                throw new ArgumentNullException(nameof(stream));
+            }
+
+            _complete = false;
+            _stream = stream;
+            _buffer = ArrayPool<byte>.Shared.Rent(1024);
```

Comment:
https://github.com/NuGet/NuGet.Client/pull/5529/files#r1415144302
> Since most filesystems now use block sizes 4k or larger, I think using a smaller buffer size may be inefficient, especially since the vast majority of NuGet's assets files will be much, much larger than 1k (or even 4k). I think STJ's JsonSerializer starts with a default buffer size of 8k, so maybe we should learn from their expertise and use that as the initial buffer size too. Even if this class ends up getting reused for other NuGet json files that are not as large as the assets file, I still think that using 4k or 8k initial buffers would be worthwhile. - @zivkan 

When I was testing this I never found us actually changing the size of the buffer. The buffer is only changed when the individual token being read or skipped exceeds the buffer size. I'm happy to adjust  the buffer size if you think it's still necessary.

Reply:
My comment isn't about resizing, it's about general performance. I have a tendency to overthink things, and be wrong about performance, so this might be wrong, but here are my thoughts:

Consider what has to happen when a program (written in C) reads 1k from a file. The app makes a syscall, so the app then transitions to kernel space, assuming the file contents aren't in the OS filesystem cache, it needs to go to the filesystem driver, the filesystem says "this 1k location is part of this 4k filesystem block", then there's lower level stuff to read the 4k block from disk, then the OS will copy the 1k chunk into the app's buffer, throwing away (well, ignoring) the other 3k. The next time the app wants to read another 1k, the whole thing happens again, although this time the file contents is likely to be in the OS filesystem cache. But it still needs a userspace to kernel transition and back again.

If the app does reads in multiples of 4k, then there will be fewer userspace-kernel transitions, and if the app is careful to do all reads starting from a 4k block boundary (block aligned reads), there will be no partial blocks read either.

With our code, it's a bit more complex, because I don't know what buffering `Stream` does. Also, to make sure that reads are aligned, that start of file, 3 byte BOM, needs to be handled carefully.

Again, I could be way off the mark here, but with my theoretical computer science knowledge, it seems to me that using a 4k or 8k buffer size might help performance, and there's pretty much no downside (honestly, the block aligned reads is probably the most important micro-optimization for perf, but I just don't know how .NET `Stream`'s own buffering affects us).

Also, it looks like I was wrong about System.Text.Json's default buffer size. They seem to use 16k: https://github.com/dotnet/runtime/blob/2b571c505a444522892a272ea7cf6ff485f59da5/src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializerOptions.cs#L24

It's a shame they didn't explain why it's that size, but perhaps they have experience we don't, so we can learn from them by using the same buffer size?  Now that we're using the array pool (the same array pool that System.Text.Json uses), then using the same buffer size will pretty much ensure that we'll share buffers with `JsonSerializer` (which NuGetAudit does for the vulnerability database) where as using a different buffer size might result at least one additional allocation.

Reply:
Okay that makes sense, I think. I've updated it to match that number, though when using ArrayPool we may in theory get something larger right?